executor/src/witgen/multiplicity_column_generator.rs

```diff
@@ -0,0 +1,208 @@
+            )
+            .collect::<BTreeMap<_, _>>();
+        let terminal_values = OwnedTerminalValues {
+            trace: all_columns,
+            public_values: publics
+                .into_iter()
+                // Publics might be unavailable if they are later-stage publics.
+                .filter_map(|(k, v)| v.map(|v| (k, v)))
+                .collect(),
+            challenge_values: self.fixed.challenges.clone(),
+        };
+
+        log::trace!(
+            "  Done building trace values, took: {}s",
+            start.elapsed().as_secs_f64()
+        );
+
+        for lookup in phantom_lookups {
+            log::trace!("  Updating multiplicity for phantom lookup: {lookup}");
+            let start = std::time::Instant::now();
+
+            let (rhs_machine_size, rhs_tuples) = self.get_tuples(&terminal_values, lookup.right);
+
+            log::trace!(
+                "    Done collecting RHS tuples, took {}s",
+                start.elapsed().as_secs_f64()
+            );
+            let start = std::time::Instant::now();
+
+            let index = rhs_tuples
+                .iter()
+                .map(|(i, tuple)| {
+                    // There might be multiple identical rows, but it's fine, we can pick any.
+                    (tuple, *i)
+                })
+                .collect::<HashMap<_, _>>();
+
+            log::trace!(
+                "    Done building index, took {}s",
+                start.elapsed().as_secs_f64()
+            );
+            let start = std::time::Instant::now();
+
+            let (_, lhs_tuples) = self.get_tuples(&terminal_values, lookup.left);
+
+            log::trace!(
+                "    Done collecting LHS tuples, took: {}s",
+                start.elapsed().as_secs_f64()
+            );
+            let start = std::time::Instant::now();
+
+            let multiplicity_column_id = lookup.multiplicity_column.unwrap();
+            let multiplicities = multiplicity_columns
+                .entry(multiplicity_column_id)
+                .or_insert_with(|| vec![0; rhs_machine_size]);
+            assert_eq!(multiplicities.len(), rhs_machine_size);
+
+            // Looking up the index is slow, so we do it in parallel.
+            let indices = lhs_tuples
+                .into_par_iter()
+                .map(|(_, tuple)| index[&tuple])
+                .collect::<Vec<_>>();
+
+            for index in indices {
+                multiplicities[index] += 1;
+            }
+            log::trace!(
+                "    Done updating multiplicities, took: {}s",
+                start.elapsed().as_secs_f64()
+            );
+        }
+
+        let columns = terminal_values
+            .destroy()
+            .into_iter()
+            .chain(
+                multiplicity_columns
+                    .into_iter()
+                    .map(|(col, values)| (col, values.into_iter().map(T::from).collect())),
```

Comment:
Aren't RHS multiplicities supposed to be negative?

Reply:
The column itself (which is what is referenced by the phantom lookup) stores the positive multiplicities, but the `bus_receive` call [here](https://github.com/powdr-labs/powdr/blob/067b63324b0f7ffca61bb3c3cc06fe798a6f2306/std/protocols/lookup_via_bus.asm#L23) multiplies it with `-1`. So I think this is correct.